server/app/scanpy_engine/scanpy_engine.py

```diff
@@ -0,0 +1,183 @@
+import os
+
+import numpy as np
+import scanpy.api as sc
+from scipy import stats
+
+from ..util.schema_parse import parse_schema
+from ..driver.driver import CXGDriver
+
+
+class ScanpyEngine(CXGDriver):
+
+    def __init__(self, data, schema=None, graph_method="umap", diffexp_method="ttest"):
+        self.data = self._load_data(data)
+        self.schema = self._load_or_infer_schema(data, schema)
+        self._set_cell_ids()
+        self.cell_count = self.data.shape[0]
+        self.gene_count = self.data.shape[1]
+        self.graph_method = graph_method
+        self.diffexp_method = diffexp_method
+
+    @staticmethod
+    def _load_data(data):
+        return sc.read(os.path.join(data, "data.h5ad"))
+
+    @staticmethod
+    def _load_or_infer_schema(data, schema):
+        data_schema = None
+        if not schema:
+            pass
+        else:
+            data_schema = parse_schema(os.path.join(data, schema))
+        return data_schema
+
+    def _set_cell_ids(self):
+        self.data.obs['cxg_cell_id'] = list(range(self.data.obs.shape[0]))
+        self.data.obs["cell_name"] = list(self.data.obs.index)
+        self.data.obs.set_index('cxg_cell_id', inplace=True)
+
+    def cells(self):
+        return list(self.data.obs.index)
+
+    def cellids(self, df=None):
+        if df:
+            return list(df.obs.index)
+        else:
+            return list(self.data.obs.index)
+
+    def genes(self):
+        return self.data.var.index.tolist()
+
+    def filter_cells(self, filter):
```

Comment:
To better understand this, I think it would be helpful to know what filter does and behaves.

Reply:
Here is how filters work: 
For categorical metadata keys filter based on key=value 
For continuous metadata keys filter by key=min,max
Either value can be replaced by a *. To have only a minimum value key=min,* To have only a maximum value key=*,max 

They combine via AND so a cell's metadata would have to match every filter

The results is a matrix with the cells the pass the filter and at this point all the genes